include/revng/ADT/GenericGraph.h

```diff
@@ -388,6 +393,277 @@ class BidirectionalNode
   NeighborContainer Predecessors;
 };
 
+template<typename T>
+concept IsBidirectionalNode = requires {
+  T::is_bidirectional_node;
+  typename llvm::Inverse<T *>;
+};
+
+namespace detail {
+/// The parameters deciding specifics of the base type `NewNode` extends are
+/// non-trivial. That's why those decisiion were wrapped inside this struct
+/// to minimize clutter.
+///
+/// \note At this step `TheNode` has not been declared yet, thus we accept a
+///       template parameter that has the same signature as `TheNode` that
+///       will be declared later. This allows us to use it as if it was
+///       declared, provided that only the real `TheNode` is used as this
+///       argument.
+template<typename Node,
+         typename EdgeLabel,
+         bool HasParent,
+         size_t SmallSize,
+         template<typename, typename, bool, size_t, typename, size_t, bool>
+         class TheNode,
+         typename FinalType,
+         size_t ParentSmallSize,
+         bool ParentHasEntryNode>
+struct NewNodeBaseTCalc {
+  static constexpr bool
+    NoDerivation = std::is_same_v<FinalType, std::false_type>;
+  using NodeType = TheNode<Node,
+                           EdgeLabel,
+                           HasParent,
+                           SmallSize,
+                           FinalType,
+                           ParentSmallSize,
+                           ParentHasEntryNode>;
+  using DerivedType = std::conditional_t<NoDerivation, NodeType, FinalType>;
+  using GenericGraph = GenericGraph<DerivedType,
+                                    ParentSmallSize,
+                                    ParentHasEntryNode>;
+  using ParentType = Parent<GenericGraph, Node>;
+  using Result = std::conditional_t<HasParent, ParentType, Node>;
+};
+} // namespace detail
+
+/// A node type with support for non-trivial edge types.
+template<typename Node,
+         typename EdgeLabel = Empty,
```

Comment:
Can we optimize the case in which there's no `EdgeLabel`?
I'd be happy to avoid dynamic allocations if possible in that case.

Reply:
Yes, that sounds like an incredible idea, how have I not thought about it, thank you!

Reply:
But wait, if so, we need to decide whether we make this `complex` node type standalone, or merge it together with a const label alternative. Because if so, empty label is implicitly const - it can't be changed if there's nothing to change. Otherwise, maybe it's a good idea to just make the `EdgeLabel` type a necessary parameter.

Reply:
Uhm, I'm not 100% sure I understand what you mean.
I was just suggesting to avoid `std::make_unique<EdgeLabel>` if `EdgeLabel` is `Empty`.

Reply:
Yes, I agree with your suggestion, what I'm thinking about, is that maybe we should explicitly disallow `NewNode<NodeData, Empty>` because it defeats the purpose of this node type.

But yeah, I'm probably overthinking this stuff again, my bad.

Reply:
I think it makes sense to keep the `Empty` option for now.

Reply:
Done.